core/drivers/imx_snvs.c

```diff
@@ -96,3 +196,27 @@ bool plat_rpmb_key_is_ready(void)
 	return ssm_secure && (security == SNVS_SECURITY_CFG_CLOSED);
 }
 #endif /* CFG_RPMB_FS */
+
+TEE_Result imx_snvs_set_master_otpmk(void)
+{
+	bool status = false;
+
+	if (is_otpmk_valid()) {
+		if (is_otpmk_selected() && is_mks_locked())
+			status = true;
+		else
+			status = set_mks_otpmk();
+	}
+
+	/*
+	 * If the board is closed, setting the OTP as the master key should
+	 * not fail.
+	 */
+	if (!status & snvs_is_device_closed()) {
+		EMSG("Device is closed");
+		EMSG("Master Key cannot be set to OTPMK");
+		return TEE_ERROR_BAD_STATE;
+	}
+
```

Comment:
is this function correct? 
When the OTPMK is not valid and the device is not closed, we are returning TEE_SUCCESS.

Reply:
Yes, this is correct.
When the device is not closed, the OTP key could be not fused yet (so all zeroes ...) triggering a key violation. So, in the open state, we simply ignore the error.
Plus, in the open state, the CAAM does not use the master key provided by the SNVS by a dummy test key instead.

Reply:
um but then, the function will return success but the master OTPMK will not have been set; the semantics will be violated making the code difficult to read IMO.

can we not return something else in that case and handle it with a comment?

Reply:
Would something like this more readable?
```suggestion

TEE_Result imx_snvs_set_master_otpmk(void)
{
	bool status = false;

	/* The OTPMK can only be set on closed devices */
	if (!snvs_is_device_closed())
		return TEE_ERROR_BAD_STATE;

	if (is_otpmk_valid())
		return TEE_ERROR_BAD_STATE;

	if (is_otpmk_selected() && is_mks_locked())
		return TEE_SUCCESS;

	if (set_mks_otpmk())
		return TEE_SUCCESS;
	else
		return TEE_ERROR_BAD_STATE;
}
```

Reply:
yes. something like that.

I personally would do something like this - and remove the static set_mks_otpmk() but what you did is the sort of thing I had in mind.

```
TEE_Result imx_snvs_set_master_otpmk(void){

	vaddr_t base = core_mmu_get_va(SNVS_BASE, MEM_AREA_IO_SEC, SNVS_SIZE);

	if (!snvs_is_device_closed())
		return TEE_ERROR_BAD_STATE;

	if (is_otpmk_valid())
		return TEE_ERROR_BAD_STATE;

	if (is_mks_locked()) {
		if (is_otpmk_selected())
			return TEE_SUCCESS;
		return TEE_ERROR_BAD_STATE;
	}

	/* set the key */
	io_setbits32(base + SNVS_HPCOMR, SNVS_HPCOMR_MKS_EN);
	io_clrbits32(base + SNVS_LPMKCR, SNVS_LPMKCR_MKCR_MKS_SEL);
	io_clrbits32(base + SNVS_HPLR, SNVS_HPLR_MKS_SL);
	io_setbits32(base + SNVS_LPLR, SNVS_LPLR_MKS_HL);

	return TEE_SUCCES;
}

```

Reply:
I'm okay with your implementation. I will just keep the register access for the key select and lock behind `set_mks_otpmk()` function.